File: course_1/module_2/examples/strassen_matrix_multiplier.py
class StrassenMatrixMultiplier:
    def __init__(self, matrix_a, matrix_b):
        # Check that matrices A and B are square and of the same dimensions.
        if not (len(matrix_a) == len(matrix_a[0]) == len(matrix_b) == len(matrix_b[0])):
            raise ValueError("Matrices must be square and of the same dimensions")

        # If the size is not a power of 2, pad matrices to the nearest power of 2.
        self.size = len(matrix_a)
        if not (self.size & (self.size - 1) == 0 and self.size != 0):  # Check if size is a power of 2.
            self.A = self._pad_to_power_of_two(matrix_a)
            self.B = self._pad_to_power_of_two(matrix_b)
        else:
            self.A = matrix_a
            self.B = matrix_b

    def multiply(self):
        # Perform Strassen's multiplication and remove padding if it was applied.
        final_result = self._strassen(self.A, self.B)
        # Remove padding if any was added.
        return [final_row[:self.size] for final_row in final_result[:self.size]]

    def _strassen(self, matrix_a, matrix_b):
        n = len(matrix_a)
        # Base case: If the matrix is 1x1, return scalar product directly.
        if n == 1:
            return [[matrix_a[0][0] * matrix_b[0][0]]]

        # Split matrices into quadrants.
        a11, a12, a21, a22 = self._split_matrix(matrix_a)
        b11, b12, b21, b22 = self._split_matrix(matrix_b)

        # Calculate the 7 products using Strassen's algorithm.
        p1 = self._strassen(a11, self._elementwise_subtract(b12, b22))
        p2 = self._strassen(self._elementwise_add(a11, a12), b22)
        p3 = self._strassen(self._elementwise_add(a21, a22), b11)
        p4 = self._strassen(a22, self._elementwise_subtract(b21, b11))
        p5 = self._strassen(self._elementwise_add(a11, a22), self._elementwise_add(b11, b22))
        p6 = self._strassen(self._elementwise_subtract(a12, a22), self._elementwise_add(b21, b22))
        p7 = self._strassen(self._elementwise_subtract(a11, a21), self._elementwise_add(b11, b12))

        # Combine the products to get the final quadrants of the resulting matrix.
        c11 = self._elementwise_add(self._elementwise_subtract(self._elementwise_add(p5, p4), p2), p6)
        c12 = self._elementwise_add(p1, p2)
        c21 = self._elementwise_add(p3, p4)
        c22 = self._elementwise_subtract(self._elementwise_subtract(self._elementwise_add(p5, p1), p3), p7)

        # Combine quadrants into a single matrix.
        return self._combine_quadrants(c11, c12, c21, c22)

    @staticmethod
    def _split_matrix(matrix):
        # Split a matrix into four quadrants.
        mid = len(matrix) // 2
        a11 = [matrix_row[:mid] for matrix_row in matrix[:mid]]
        a12 = [matrix_row[mid:] for matrix_row in matrix[:mid]]
        a21 = [matrix_row[:mid] for matrix_row in matrix[mid:]]
        a22 = [matrix_row[mid:] for matrix_row in matrix[mid:]]
        return a11, a12, a21, a22

    @staticmethod
    def _combine_quadrants(c11, c12, c21, c22):
        # Combine four quadrants into a single matrix.
        top = [row_c11 + row_c12 for row_c11, row_c12 in zip(c11, c12)]
        bottom = [row_c21 + row_c22 for row_c21, row_c22 in zip(c21, c22)]
        return top + bottom

    @staticmethod
    def _pad_to_power_of_two(matrix):
        """Pad the matrix with zeros to the nearest power of two size."""
        n = len(matrix)
        new_size = 1 << (n - 1).bit_length()
        padded_matrix = [[0] * new_size for _ in range(new_size)]
        for i in range(n):
            for j in range(n):
                padded_matrix[i][j] = matrix[i][j]
        return padded_matrix

    @staticmethod
    def _elementwise_add(matrix_a, matrix_b):
        # Add two matrices element-wise.
        n = len(matrix_a)
        return [[matrix_a[i][j] + matrix_b[i][j] for j in range(n)] for i in range(n)]

    @staticmethod
    def _elementwise_subtract(matrix_a, matrix_b):
        # Subtract matrix_b from matrix_a element-wise.
        n = len(matrix_a)
        return [[matrix_a[i][j] - matrix_b[i][j] for j in range(n)] for i in range(n)]


# Example usage.
if __name__ == "__main__":
    # Define two square matrices of arbitrary size.
    A = [[1, 2], [3, 4]]
    B = [[5, 6], [7, 8]]

    multiplier = StrassenMatrixMultiplier(A, B)
    result = multiplier.multiply()

    print("Matrix A:")
    for row in A:
        print(row)
    print("Matrix B:")
    for row in B:
        print(row)
    print("Result of Strassen's Matrix Multiplication (A * B):")
    for row in result:
        print(row)

File: course_1/module_2/examples/strassen_matrix_multiplier.go
package main

import (
	"errors"
	"fmt"
)

// StrassenMatrixMultiplier is a struct for matrix multiplication using Strassen's algorithm.
type StrassenMatrixMultiplier struct {
	MatrixA, MatrixB [][]int
	AdjustedSize     int
	OriginalSize     int
}

// NewStrassenMatrixMultiplier is a constructor that initializes a StrassenMatrixMultiplier instance.
func NewStrassenMatrixMultiplier(matrixA, matrixB [][]int) (*StrassenMatrixMultiplier, error) {
	// Ensure matrices are square and of the same dimensions.
	if len(matrixA) != len(matrixB) || len(matrixA) != len(matrixA[0]) || len(matrixB) != len(matrixB[0]) {
		return nil, errors.New("both matrices must be square and have the same dimensions")
	}

	originalSize := len(matrixA)
	adjustedSize := getNextPowerOfTwo(originalSize)

	// Pad matrices to the nearest power of two if necessary.
	if adjustedSize != originalSize {
		matrixA = padMatrix(matrixA, adjustedSize)
		matrixB = padMatrix(matrixB, adjustedSize)
	}

	return &StrassenMatrixMultiplier{
		MatrixA:      matrixA,
		MatrixB:      matrixB,
		AdjustedSize: adjustedSize,
		OriginalSize: originalSize,
	}, nil
}

// Multiply performs Strassen's matrix multiplication and removes padding if added.
func (s *StrassenMatrixMultiplier) Multiply() [][]int {
	result := s.strassenMultiply(s.MatrixA, s.MatrixB)
	// Remove padding if it was added to the matrices
	return unPadMatrix(result, s.OriginalSize)
}

// strassenMultiply is a recursive method for performing Strassen's matrix multiplication.
func (s *StrassenMatrixMultiplier) strassenMultiply(matrixA, matrixB [][]int) [][]int {
	size := len(matrixA)
	// Base case: If the matrix is 1x1, return the product directly.
	if size == 1 {
		return [][]int{{matrixA[0][0] * matrixB[0][0]}}
	}

	// Split matrices into quadrants.
	a11, a12, a21, a22 := splitMatrix(matrixA)
	b11, b12, b21, b22 := splitMatrix(matrixB)

	// Compute the 7 products required by Strassen's algorithm.
	product1 := s.strassenMultiply(a11, subtractMatrix(b12, b22))
	product2 := s.strassenMultiply(addMatrix(a11, a12), b22)
	product3 := s.strassenMultiply(addMatrix(a21, a22), b11)
	product4 := s.strassenMultiply(a22, subtractMatrix(b21, b11))
	product5 := s.strassenMultiply(addMatrix(a11, a22), addMatrix(b11, b22))
	product6 := s.strassenMultiply(subtractMatrix(a12, a22), addMatrix(b21, b22))
	product7 := s.strassenMultiply(subtractMatrix(a11, a21), addMatrix(b11, b12))

	// Combine products to form the resulting quadrants.
	c11 := addMatrix(subtractMatrix(addMatrix(product5, product4), product2), product6)
	c12 := addMatrix(product1, product2)
	c21 := addMatrix(product3, product4)
	c22 := subtractMatrix(subtractMatrix(addMatrix(product5, product1), product3), product7)

	// Combine quadrants into a single matrix.
	return combineQuadrants(c11, c12, c21, c22)
}

// getNextPowerOfTwo finds the smallest power of two greater than or equal to the given integer.
func getNextPowerOfTwo(n int) int {
	power := 1
	for power < n {
		power *= 2
	}
	return power
}

// padMatrix expands the matrix to the specified size, filling in zeros for the new entries.
func padMatrix(matrix [][]int, size int) [][]int {
	padded := make([][]int, size)
	for i := range padded {
		padded[i] = make([]int, size)
		if i < len(matrix) {
			copy(padded[i], matrix[i])
		}
	}
	return padded
}

// unPadMatrix removes extra padding from the matrix, restoring it to the original size.
func unPadMatrix(matrix [][]int, size int) [][]int {
	unPadded := make([][]int, size)
	for i := 0; i < size; i++ {
		unPadded[i] = matrix[i][:size]
	}
	return unPadded
}

// splitMatrix divides a matrix into four quadrants.
func splitMatrix(matrix [][]int) ([][]int, [][]int, [][]int, [][]int) {
	halfSize := len(matrix) / 2
	a11 := make([][]int, halfSize)
	a12 := make([][]int, halfSize)
	a21 := make([][]int, halfSize)
	a22 := make([][]int, halfSize)
	for i := 0; i < halfSize; i++ {
		a11[i] = matrix[i][:halfSize]
		a12[i] = matrix[i][halfSize:]
		a21[i] = matrix[i+halfSize][:halfSize]
		a22[i] = matrix[i+halfSize][halfSize:]
	}
	return a11, a12, a21, a22
}

// combineQuadrants merges four quadrants into a single matrix.
func combineQuadrants(c11, c12, c21, c22 [][]int) [][]int {
	size := len(c11) * 2
	combined := make([][]int, size)
	for i := 0; i < size/2; i++ {
		combined[i] = append(c11[i], c12[i]...)
		combined[i+size/2] = append(c21[i], c22[i]...)
	}
	return combined
}

// addMatrix adds two matrices element-wise.
func addMatrix(matrixA, matrixB [][]int) [][]int {
	size := len(matrixA)
	result := make([][]int, size)
	for i := 0; i < size; i++ {
		result[i] = make([]int, size)
		for j := 0; j < size; j++ {
			result[i][j] = matrixA[i][j] + matrixB[i][j]
		}
	}
	return result
}

// subtractMatrix subtracts matrixB from matrixA element-wise.
func subtractMatrix(matrixA, matrixB [][]int) [][]int {
	size := len(matrixA)
	result := make([][]int, size)
	for i := 0; i < size; i++ {
		result[i] = make([]int, size)
		for j := 0; j < size; j++ {
			result[i][j] = matrixA[i][j] - matrixB[i][j]
		}
	}
	return result
}

// Main function for testing.
func main() {
	matrixA := [][]int{
		{1, 2},
		{3, 4},
	}
	matrixB := [][]int{
		{5, 6},
		{7, 8},
	}

	multiplier, err := NewStrassenMatrixMultiplier(matrixA, matrixB)
	if err != nil {
		fmt.Println("Error:", err)
		return
	}

	result := multiplier.Multiply()

	fmt.Println("Matrix A:")
	printMatrix(matrixA)
	fmt.Println("Matrix B:")
	printMatrix(matrixB)
	fmt.Println("Result of Strassen's Matrix Multiplication (A * B):")
	printMatrix(result)
}

// printMatrix prints a matrix row by row.
func printMatrix(matrix [][]int) {
	for _, row := range matrix {
		fmt.Println(row)
	}
}
